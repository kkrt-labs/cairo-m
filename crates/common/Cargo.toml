[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
