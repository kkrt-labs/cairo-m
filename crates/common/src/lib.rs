pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
