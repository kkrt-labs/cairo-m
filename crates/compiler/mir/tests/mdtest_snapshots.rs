//! MIR snapshot tests for mdtest cases.
//! This file automatically generates MIR snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the MIR generation pipeline.

mod common;

use cairo_m_compiler_mir::{generate_mir, PrettyPrint};
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_mir_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("MIR", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            match generate_mir(&db, crate_id) {
                Ok(module) => Ok(module.pretty_print(0)),
                Err(diagnostics) => Err(format!(
                    "MIR generation failed with diagnostics:\n{:#?}",
                    diagnostics
                )),
            }
        });

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("MIR snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
