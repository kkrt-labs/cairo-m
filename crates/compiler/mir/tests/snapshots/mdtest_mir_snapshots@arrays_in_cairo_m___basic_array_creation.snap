---
source: crates/compiler/mir/tests/mdtest_snapshots.rs
description: "MIR snapshot for mdtest: Arrays in Cairo-M - Basic Array Creation"
input_file: mdtest/01-basics/05-arrays.md
---
Source:
fn create_array() -> felt {
    let arr: [felt; 5] = [1, 2, 3, 0, 0];
    return arr[4];
}
============================================================
Generated MIR:
module {
  // Function 0
  fn create_array {
    entry: 0

    0:
      %0 = makefixedarray [1, 2, 3, 0, 0]
      %1 = arrayindex %0, 4
      return %1

  }

}
